File: Source/AlphaTab/Rendering/Glyphs/TabRestGlyph.cs
using AlphaTab.Model;
using AlphaTab.Platform;
using AlphaTab.Rendering.Utils;

namespace AlphaTab.Rendering.Glyphs
{
    class TabRestGlyph : MusicFontGlyph
    {
        private readonly bool _isVisibleRest;
        private readonly Duration _duration;

        public BeamingHelper BeamingHelper { get; set; }

        public TabRestGlyph(float x, float y, bool isVisibleRest, Duration duration)
            : base(x, y, 1, ScoreRestGlyph.GetSymbol(duration))
        {
            _isVisibleRest = isVisibleRest;
            _duration = duration;
        }

        public override void DoLayout()
        {
            if (_isVisibleRest)
            {
                Width = ScoreRestGlyph.GetSize(_duration) * Scale;
            }
            else
            {
                Width = 10 * Scale;
            }
        }

        public void UpdateBeamingHelper(float cx)
        {
            if (BeamingHelper != null && BeamingHelper.IsPositionFrom(TabBarRenderer.StaffId, Beat))
            {
                BeamingHelper.RegisterBeatLineX(TabBarRenderer.StaffId, Beat, cx + X + Width, cx + X);
            }
        }

        public override void Paint(float cx, float cy, ICanvas canvas)
        {
            if (_isVisibleRest)
            {
                base.Paint(cx, cy, canvas);
            }
        }
    }
}

File: src/rendering/glyphs/TabRestGlyph.ts
import { BeatSubElement } from '@src/model/Beat';
import type { Duration } from '@src/model/Duration';
import type { ICanvas } from '@src/platform/ICanvas';
import { MusicFontGlyph } from '@src/rendering/glyphs/MusicFontGlyph';
import { ScoreRestGlyph } from '@src/rendering/glyphs/ScoreRestGlyph';
import type { BeamingHelper } from '@src/rendering/utils/BeamingHelper';
import { ElementStyleHelper } from '@src/rendering/utils/ElementStyleHelper';

export class TabRestGlyph extends MusicFontGlyph {
    private _isVisibleRest: boolean;
    public beamingHelper!: BeamingHelper;

    public constructor(x: number, y: number, isVisibleRest: boolean, duration: Duration) {
        super(x, y, 1, ScoreRestGlyph.getSymbol(duration));
        this._isVisibleRest = isVisibleRest;
    }

    public override doLayout(): void {
        super.doLayout();
        if (!this._isVisibleRest) {
            this.width = 10;
        }
    }

    public updateBeamingHelper(cx: number): void {
        if (this.beamingHelper && this.beamingHelper.isPositionFrom('tab', this.beat!)) {
            this.beamingHelper.registerBeatLineX(
                'tab',
                this.beat!,
                cx + this.x + this.width / 2,
                cx + this.x + this.width / 2
            );
        }
    }

    public override paint(cx: number, cy: number, canvas: ICanvas): void {
        if (this._isVisibleRest) {
            using _ = ElementStyleHelper.beat(canvas, BeatSubElement.GuitarTabRests, this.beat!);
            super.paint(cx, cy, canvas);
        }
    }
}
